Add resource v1beta2 API

mortent opened the pull request:
#### What type of PR is this?

/kind api-change
/kind deprecation

#### What this PR does / why we need it:

This adds the new v1beta2 API for resource.k8s.io as described in https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4816-dra-prioritized-list. This also flattens the BasicDevice type in the ResourceSlice and updates the nodeSelection fields on the ResourceSliceSpec to be pointers (ref https://github.com/kubernetes/kubernetes/issues/130932)

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/130932

#### Special notes for your reviewer:

The prioritized list feature that is described in the KEP is implemented in https://github.com/kubernetes/kubernetes/pull/128586.

#### Does this PR introduce a user-facing change?
```release-note
The resource.k8s.io/v1beta1 API is deprecated and will be removed in 1.36. Use v1beta2 instead.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:
- [KEP] https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4816-dra-prioritized-list

```docs

```

mortent commented:
/test pull-kubernetes-kind-dra-all

mortent commented:
This will be updated once https://github.com/kubernetes/kubernetes/pull/128586 has merged

pohly commented:
Actually...

/milestone v1.33

cici37 commented:
/assign

mortent commented:
/test pull-kubernetes-kind-dra-all

tico88612 commented:
Hello, @cici37 @mortent 
Appreciate all of your efforts with this PR! Is the plan still to resolve it for v1.33 release?
If so, a gentle reminder that the code freeze has started [02:00 UTC Friday 21st March 2025](https://everytimezone.com/s/2c5e9275) . Please make sure any PRs have both lgtm and approved labels ASAP, and file an [Exception](https://github.com/kubernetes/sig-release/blob/master/releases/EXCEPTIONS.md) if you haven't done it yet.
Thanks!

mortent commented:
@tico88612 Yeah, we still plan to get this in before the freeze. But it is currently blocked by https://github.com/kubernetes/kubernetes/pull/130764.

pohly commented:
This fixes https://github.com/kubernetes/kubernetes/issues/130932

mortent committed:
Add resource v1beta2 API

cici37 commented on pkg/printers/internalversion/printers.go:
```diff
@@ -3171,7 +3171,11 @@ func printResourceSlice(obj *resource.ResourceSlice, options printers.GenerateOp
+	if obj.Spec.NodeName != nil {
+		nodeName = *obj.Spec.NodeName
+	}
+	row.Cells = append(row.Cells, obj.Name, nodeName, obj.Spec.Driver, obj.Spec.Pool.Name, translateTimestampSince(obj.CreationTimestamp))
```
nit: I am guessing we only add nodeName into the printer values because it's immutable? Still feel a little bit off by only print one field from a one-of fields list but can live with it :)

cici37 reviewed (commented):
Hold the review until field name change sorted out :)

mortent commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -203,10 +185,19 @@ func validateDeviceRequest(request resource.DeviceRequest, fldPath *field.Path,
+		allErrs = append(allErrs, field.Required(fldPath, "exactly one of `exactly` or `firstAvailable` is required"))
+	case 1:
+	default:
+		allErrs = append(allErrs, field.Invalid(fldPath, nil, "exactly one of `exactly` or `firstAvailable` is required"))
```
Fixed

mortent commented on pkg/apis/resource/validation/validation.go:
```diff
@@ -587,9 +589,13 @@ func validateResourceSliceSpec(spec, oldSpec *resource.ResourceSliceSpec, fldPat
 
 	setFields := make([]string, 0, 4)
-	if spec.NodeName != "" {
+	if spec.NodeName != nil {
```
Yeah, I've updated the validation logic for `NodeName` to match what we do for the boolean fields.

mortent commented:
/test pull-kubernetes-unit
/test pull-kubernetes-e2e-gce

I think these are both flakes. Will follow up on the DRA failure in https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129970/pull-kubernetes-unit/1902774424365436928 tomorrow.

Jefftree commented:
/remove-sig api-machinery

pohly commented:
> Will follow up on the DRA failure in https://prow.k8s.io/view/gs/kubernetes-ci-logs/pr-logs/pull/129970/pull-kubernetes-unit/1902774424365436928 tomorrow.

Fixed by https://github.com/kubernetes/kubernetes/pull/130947.

pohly commented on pkg/apis/resource/validation/validation.go:
@cici37, in https://github.com/kubernetes/kubernetes/pull/130764 we talked about how the size of the new nested maps needs to be constrained. I don't find the thread anymore, but I was suggested that for something that is basically a map to a map, the *total number of leaf entries* should be restricted. 

Furthermore, for a Device, the total number of attributes, quantities, and counters should be limited to `ResourceSliceMaxAttributesAndCapacitiesPerDevice = 32` (then needs to be renamed) to keep the overall size in reasonable bounds. Allowing `Device` to become larger in the worst case can make a ResourceSlice too large. With the combined limit, driver developers can decide whether they spend their "budget" more on many attributes or more on the others. If they run into the limit, they have to use mixins.

I didn't have time to check again before the PR merged, so let me bring up now: that's not how it got implemented.

https://github.com/kubernetes/kubernetes/blob/e6d2dfdd2eb24dcc041821ed066c35b28c099ede/pkg/apis/resource/validation/validation.go#L711-L715

https://github.com/kubernetes/kubernetes/blob/e6d2dfdd2eb24dcc041821ed066c35b28c099ede/pkg/apis/resource/validation/validation.go#L775-L776

With that, `ConsumesCounter` may reference `ResourceSliceMaxDeviceCounterConsumptionCounters * ResourceSliceMaxDeviceCounterConsumptions` counters, which is not how that is defined (or it is defined ambiguously):

https://github.com/kubernetes/kubernetes/blob/e6d2dfdd2eb24dcc041821ed066c35b28c099ede/pkg/apis/resource/validation/validation.go#L775-L776

I've not checked the validation of the other fields.

Validation is very hard to fix later. Can we address this in this PR?

Another comment that didn't get resolved:

https://github.com/kubernetes/kubernetes/blob/e6d2dfdd2eb24dcc041821ed066c35b28c099ede/pkg/apis/resource/validation/validation.go#L772

I prefer `len(deviceCounterConsumption.Counters) == 0` to check for an empty map.

johnbelamaric replied:
Thank you Patrick, this sounds wise to address now.

cici37 replied:
ohh I added the validation based on the original KEP which only says `// The maximum number of device capacity consumption entries is 32.` and ` // The maximum number of capacities is 32.` without aware of the total number constraints. We might need to update the API doc together with the validation. Thanks for catching it. Let me know if any help is needed for fixing :)

pohly replied:
My update of validation is in https://github.com/mortent/kubernetes/pull/2.

mortent replied:
Thanks Patrick. I've included your commit in the branch.

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+	Exactly *SpecificDeviceRequest `json:"exactly,omitempty" protobuf:"bytes,2,name=exactly"`
+
+	// FirstAvailable contains subrequests, of which exactly one will be
+	// satisfied by the scheduler to satisfy this request. It tries to
```
nit: exactly one will be selected byt he scheduler

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+	// FirstAvailable contains subrequests, of which exactly one will be
+	// satisfied by the scheduler to satisfy this request. It tries to
+	// satisfy them in the order in which they are listed here. So if
+	// there are two entries in the list, the schduler will only check
```
typo: schduler

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+	// additional configuration and selectors to be inherited by this
+	// request.
+	//
+	// A DeviceClassName is required. Clients must check that it is
```
It's a little unusual to tell clients to check a required field.

pohly replied:
Can be removed now. This was the "we reserve the right to change this" provision in previous versions.

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+	//   Allocation will fail if some devices are already allocated,
+	//   unless adminAccess is requested.
+	//
+	// If AlloctionMode is not specified, the default mode is ExactCount. If
```
typo: Alloction

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+}
+
+// SpecificDeviceRequest is a request for one or more identical devices.
+type SpecificDeviceRequest struct {
```
nit: I would call this `ExactDeviceRequest`

mortent replied:
I've renamed `SpecifiedDeviceRequest` to `ExactDeviceRequest`.

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+	// Requests lists the names of requests where the configuration applies.
+	// If empty, it applies to all requests.
+	//
+	// References to subrequests must include the name of both the main request
```
This now says "must include ...both the main request and the subrequest" and then says "if just the main request is given..." -- those are contradictory

mortent replied:
So the first  part "must include ..both the main request and the subrequest" describes references to subrequests. But agree it could be clearer.

There were a few diffs between the v1beta2 types and the internal types that I have cleaned up now and this was one of those. So the language here has been updated.

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+	// If empty, it applies to all requests.
+	//
+	// References to subrequests must include the name of both the main request
+	// and the subrequest using the format <main request>/<subrequest>. if just
```
s/if/If/

thockin commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+	// Requests lists the names of requests where the configuration applies.
+	// If empty, its applies to all requests.
+	//
+	// References to subrequests must include the name of both the main request
```
This now says "must include ...both the main request and the subrequest" and then says "if just the main request is given..." -- those are contradictory

thockin commented on pkg/apis/resource/v1beta1/conversion.go:
```diff
@@ -38,3 +43,307 @@ func addConversionFuncs(scheme *runtime.Scheme) error {
+}
+
+func Convert_v1beta1_ResourceSliceSpec_To_resource_ResourceSliceSpec(in *resourcev1beta1.ResourceSliceSpec, out *resource.ResourceSliceSpec, s conversion.Scope) error {
+	out.Driver = in.Driver
```
The "normal" pattern for manual conversion is to call the generated code (`autoConvert_v1beta1_ResourceSliceSpec_To_resource_ResourceSliceSpec`) first, then fix anything required.  That way, this file highlight only the things that are NOT automatic.

A lot of the code here seems to be identical to the autoConvert, which presents a liability - extra LOCs, tests, and risk in the face of evolution (since manual conversion is pretty rare).

I won't repeat the point, but all of these are the same comment :)

mortent replied:
Thanks for the guidance here. So for the `ResourceSliceSpec`, it seems like the autogenerated conversion handles the "type to pointer-to-type" conversion automatically, so I've removed manual conversion for that type.

For the the `DeviceRequest` and `Device` conversions, I've updated the logic to first call the `autoConvert` version. But since we have moved some of the fields around, the auto-conversion logic is missing quite a bit, so there is still manual logic required for those. I can't see any obvious ways to avoid it.

mortent replied:
So I had to add back the conversion logic for the `ResourceSliceSpec`, as the default conversion between string and string pointer didn't work quite the way I wanted it to. But the new implementation uses the autoConvert function and only fixes the conversion for the `NodeName` and `AllNodes` fields.

The logic for string conversion is:
* **string pointer to string**: nil pointer and pointer to empty string is converted to empty string (so some information is lost here)
* **string to string pointer**: empty string is converted to nil pointer.

Added validation so pointer to empty string is not allowed.

The logic for boolean conversion is:
* **bool pointer to bool**: nil pointer and pointer to `false` is converted to `false`.
* **bool to bool pointer**: `false` is converted to nil pointer.

Added validation so pointer to `false` is not allowed.

pohly commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1510 @@
+// DeviceCounterConsumption defines a set of counters that
+// a device will consume from a CounterSet.
+type DeviceCounterConsumption struct {
+	// SharedCounter defines the shared counter from which the
+	// counters defined will be consumed.
+	//
+	// +required
+	SharedCounter string `json:"sharedCounter" protobuf:"bytes,1,opt,name=sharedCounter"`
```
This is the name of a CounterSet, not of a "shared counter".

A shared counter is a `Counter` instance, which is identified by the name of its set and the counter name.

```suggestion
	// CounterSet defines the set from which the
	// counters defined will be consumed.
	//
	// +required
	CounterSet string `json:"counterSet" protobuf:"bytes,1,opt,name=counterSet"`
```

cc @cici37 @johnbelamaric

pohly replied:
I stumbled over this when updating the validation code.

Populating `SharedCounters`:

https://github.com/kubernetes/kubernetes/blob/9c2cbda9d8d0e2e6eb43b58b2369b7be90bfe564/pkg/apis/resource/validation/validation_resourceslice_test.go#L425-L430

Referencing one counter in it:

https://github.com/kubernetes/kubernetes/blob/9c2cbda9d8d0e2e6eb43b58b2369b7be90bfe564/pkg/apis/resource/validation/validation_resourceslice_test.go#L447-L457

The `SharedCounter: fmt.Sprintf("set-%d", i)` line isn't right.

pohly replied:
No, the line is right! It just looks extremely odd because of the `SharedCounter` field name. That should be `CounterSet: fmt.Sprintf("set-%d", i)`.

> So it is a bit unclear is the plural in the SharedCounters name references the multiple CounterSets or the multiple Counters.

It's the "multiple counters". That's also how the size limit is now defined: the total number of counters in that field, not the number of different sets.

pohly replied:
I suggested earlier to make it a flat list with `ShareCounters []Counter`, with the name inside the struct, but @johnbelamaric convinced me that the two-level nesting with sets as outer layer is needed for mixins.

mortent replied:
Yeah, I know. I was trying to do to many things at once and realized I had misunderstood what I was reading..

mortent replied:
> It's the "multiple counters". That's also how the size limit is now defined: the total number of counters in that field, not the number of different sets.

Yeah, I think that makes sense. I will update the API to use the plural for `ConsumesCounter` and `SharedCounter` to `CounterSet`. I think we will just continue using `consumes` as I think the take-away from the discussion on the issue is that the alternatives aren't clearly better.

pohly commented on staging/src/k8s.io/api/resource/v1beta2/types.go:
```diff
@@ -0,0 +1,1560 @@
+
+	// NodeSelector defines the nodes where the device is available.
+	//
+	// Must use exactly one term.
```
The corresponding field in v1beta1 lacks this comment. Can you add it there?

mortent replied:
Fixed

pohly reviewed (commented):
The internal types.go and the v1beta2 are now mostly in sync, but not quite. I'm not going to list them all, you can do that with
```
diff -c pkg/apis/resource/types.go <(sed -e 's;`json:",inline.*;// inline;' -e 's;metav1.TypeMeta // inline;metav1.TypeMeta;' -e 's; `json.*;;' -e 's/v1\./core./' -e's/metacore./metav1./' -e 's/package v1beta1/package resource/' staging/src/k8s.io/api/resource/v1beta2/types.go | grep -v -e '+genclient' -e '+k8s:prerelease-lifecycle-gen' )
```

v1beta1 and v1beta2 now have more or less only the expected diff, with [one exception](https://github.com/kubernetes/kubernetes/pull/129970#pullrequestreview-2706635861).

The `SharedCounter` field name in `SharedCounterConsumption` still seems [wrong to me](https://github.com/kubernetes/kubernetes/pull/129970#pullrequestreview-2705200242).

mortent commented:
The PR has been updated with:
* `ConsumesCounter` has been renamed to `ConsumesCounters`
* `SharedCounter` has been renamed to `CounterSet`
* The diffs between the internal and the `v1beta2` types have been fixed.

thockin commented:
I think all my concerns are addressed.

/approve for API

johnbelamaric commented on pkg/apis/resource/types.go:
```diff
@@ -183,7 +183,7 @@ type CounterSet struct {
 	// The name of each counter must be unique in that set and must be a DNS label.
 	//
-	// The maximum number of counters is 32.
+	// The maximum number of counters in all sets is 32.
```
Are we sure this is enough, in particular for the NVIDIA MIG use case? I think there are 10-12 individual counters there. If we limit this to 32, then we are forcing a ResourceSlice per GPU. I think that's what @klueska intended anyway. So this is probably enough until we implement Mixins.

pohly replied:
@elezar: can you provide a YAML with a "real" ResourceSlice which uses the API from this PR? It needs to include all attributes and quantities of some actual hardware. I think @klueska had a tool. Otherwise start with https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4815-dra-partitionable-devices#extending-device-and-capacity-pool-with-a-set-of-mixins? But I am not sure whether it is complete or a simplified version.

With such a YAML, we can verify that it passes validation. Might also be useful for benchmarking.

mortent replied:
Yeah, I think the total number of counters should be sufficient here. From a quick look at the DGXA100 example in the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4815-dra-partitionable-devices#putting-it-all-together-for-the-mig-use-case), it seems like we have 15 different counters:
* copy-engines
* decoders
* encoders
* jpeg-engines
* memory
* multiprocessors
* ofa-engines
* memorySlice0
* memorySlice1
* memorySlice2
* memorySlice3
* memorySlice4
* memorySlice5
* memorySlice6
* memorySlice7

I also have the understanding from @klueska that there would be a ResourceSlice per GPU.

elezar replied:
I am AFK for this week. I may be able to look into it at some point, but @klueska would be a better source of information.

klueska replied:
This was the tool I had to generate the specs from the original API design:
https://github.com/kubernetes-sigs/wg-device-management/pull/38

Unfortunately, I'm in the same boat as Evan in terms of time to look at this in more detail. I will be in the car most of today driving and then in an airplane all day tomorrow.

I haven't looked at the updated design at all, but from browsing the comments here -- it seems that mixins have been pushed out, and `CapacitiyPool`s have now been rebranded as `CounterSet`s?

The "counters" listed above are the full set for MIG devices (15 in total). I don't see this changing anytime soon (if ever).

If `CapacitiyPool`s (a.k.a. `CounterSet`s?) cannot be shared across resource slices, we will need to represent 1 full GPU + each of its possible MIG devices in a single slice. At the moment, that is a total of 26 devices, but it may increase in the future if more MIG profile are added for specialized use-cases.

In terms of attributes / capacities, I currently have 11 attributes and 1 capacity per device.  However, there have been requests for more attributes to be added. I could imagine having as many as 20 by the time to do our first official release of all this.

johnbelamaric replied:
Thanks Kevin, these numbers are helpful. Yes `CapacityPool` => `CounterSet`, mixins had to push out to make the date. CounterSet (and later Mixins) cannot be shared across slices today, but I do think eventually we will want to do that.

mortent replied:
Thanks Kevin. I have looked at the output from my updated version of the tool to generate the specs and I got the numbers listed in https://github.com/kubernetes/kubernetes/pull/129970#discussion_r2008736094. It ends up with quite a few more attributes+capacities than you mentioned, so if we support those numbers, it sounds like we should be ok for now. But yeah, we should look at removing some of the limitations around the number of MIG devices for 1.34 (at least get the process started if it requires multiple releases).

I am working on updating the spec gen tool to reflect the updates to the design after the KEP.

klueska replied:
It looks like out got 40 in https://github.com/kubernetes/kubernetes/pull/129970#discussion_r2008736094 because the set of `counters` are duplicated in the `capacities`. Remind me why this is necessary again?

With this requirement (and the expectation to have up to 20 attributes / capacities that don't explicitly include duplicates of the counters), I would now expect a total of 20 + 15*2 = 50.

mortent replied:
It is not a requirement and I don't think it is a typical scenario where those will match. This might be a result of not knowing exactly what should be exposed as capacities when we changed to the capacity pool design, since the previous design didn't distinguish between capacity consumed and advertised.

What is the one capacity you expect to expose per device?

pohly replied:
Did we settle this discussion? We can still bump this up to 64.

Because it's a once-per-slice field, the effect on the slice max size is minimal.

pohly replied:
Would mixins help if 32 is too limiting?

I can't tell whether they would apply here.

mortent replied:
So with the current design we allow for 32 attributes+capacity and then another 32 counters (assuming the number of devices is kept low enough to avoid the ResourceSlice-wide limit). So the combination is 64, which is above the ask of 50, although it comes with the caveat that we have the 1024 limit on the number of device counters in a slice. But that should be sufficient for the example, which I think had 26 DRA devices for each physical GPU.

I think mixins should help quite a bit here, since the example in the KEP uses them extensively and reduces the actual attributes/capacity/counters defined in each device to almost none. But mixins won't be available until 1.34, so I think the question is whether the current limits will sufficient (but maybe not ideal) until then. We have confirmed that it is sufficient for the example in the KEP and from what I can tell, it does have some extra room for both attributes/capacity and counters.

But it would be great if @klueska or someone else can confirm.

pohly replied:
Okay, so you and @klueska were talking about the limitation on counters referenced in  devices. That belongs into https://github.com/kubernetes/kubernetes/pull/129970#discussion_r2008408292.

This thread here is about `slice.spec.sharedCounters`. I agree that we seem to be fine with the current 32 limit, but it wouldn't be hard to raise it either if someone prefers to have more flexibility.

pohly replied:
@johnbelamaric @klueska: any last comments on this?

Ready to lift the hold?

johnbelamaric commented on pkg/apis/resource/types.go:
```diff
@@ -251,9 +250,9 @@ const ResourceSliceMaxDeviceCounterConsumptions = 32
 const ResourceSliceMaxSharedCountersCounters = 32
```
Do we still need this?

mortent replied:
There were a few constants that I added back by mistake. I've cleaned that up now.

pohly commented:
/hold

To ensure that we don't accidentally merge after an LGTM.

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -322,11 +341,11 @@ type Device struct {
 // a device will consume from a CounterSet.
 type DeviceCounterConsumption struct {
-	// SharedCounter defines the shared counter from which the
+	// CounterSet defines the set from which the
```
Nit:
"CounterSet is the name of the set from which..."

Below:

"Counters defines the counters that will be consumed by the device."

mortent replied:
Fixed

mortent commented on pkg/apis/resource/types.go:
```diff
@@ -265,45 +264,34 @@ type Device struct {
 	// The name of each attribute must be unique in that set.
 	//
-	// The maximum number of attributes and capacities combined is 32.
+	// The maximum number of attributes, capacities and counters combined is 32.
```
I've done a quick check here, and it seems like 32 is not enough. When I expand the mixins in the DGXA100 example from the [KEP](https://github.com/kubernetes/enhancements/tree/master/keps/sig-scheduling/4815-dra-partitionable-devices#putting-it-all-together-for-the-mig-use-case), I end up with a total of:
* 15 counters:
  * copy-engines
  * decoders
  * encoders
  * jpeg-engines
  * memory
  * multiprocessors
  * ofa-engines
  * memorySlice0
  * memorySlice1
  * memorySlice2
  * memorySlice3
  * memorySlice4
  * memorySlice5
  * memorySlice6
  * memorySlice7
* 10 attributes:
  * cudaDriverVersion
  * driverVersion
  * architecture
  * brand
  * cudaComputeCapability
  * productName
  * type
  * index
  * minor
  * uuid
* 15 capacities
  * copy-engines
  * decoders
  * encoders
  * jpeg-engines
  * memory
  * multiprocessors
  * ofa-engines
  * memorySlice0
  * memorySlice1
  * memorySlice2
  * memorySlice3
  * memorySlice4
  * memorySlice5
  * memorySlice6
  * memorySlice7

This adds up to 40. I will update the generator in https://github.com/mortent/wg-device-management/tree/PartitionableDevicesGen with the new names and the lack of mixins. That also gives a fully expanded output. But I might not have time to do this until tomorrow.

johnbelamaric replied:
Thanks. We also have the option of implementing IntRange for counters now that they are a separate type. That would compress memory slices from 8 to 1 counter.

pohly replied:
Let's write a test case in validation.go:
- Generate a ResourceSlice that maxes out all limits for all fields (maximum number of entries, maximum length for strings, etc.), choosing the worst case scenario whenever there alternatives. For example, a string attribute is longer than an integer attribute.
- Validate it: must pass.
- Check the result of `Size()`: that's the byte-accurate size of the protobuf encoding. It should stay under 1MB.

Then we know whether we have some buffer left to perhaps add 16 counters alongside 32 attributes+capacities.

If not, then we need to become creative: if counters are used, allow less devices, for example.

mortent replied:
I've added a test that created a (at least close to) "worst-case" `ResourceSliceSpec`, validates it, and then checks the size. It comes in at just over 1MB. I think checking the size of just the spec might be better than the overall ResourceSlice, since it contains many fields that are unrelated to DRA.

Given that we already know that other fields are coming (mixins, ...), we probably don't want to add additional room for more counters. Limiting the number of devices when other features are in use might be our best option.

Also found that we lacked validation on the number of traits on a device, so added that and a test.

pohly replied:
> I think checking the size of just the spec might be better than the overall ResourceSlice, since it contains many fields that are unrelated to DRA.

That's part of the problem: we have to leave room for metadata. So we have to look at the overall size.

I was thinking about this over the weekend: just doing a validation check is not enough. We need to do at least one real CREATE with an apiserver and then look at the result, because then we also get the mandatory ManagedFields in the metadata. That can be potentially large. We avoid the worst of it by not using `+listType=map`, but that might not be enough.

I'll have a look at your test.

mortent replied:
So I agree that in the end it is the total size that matters, but we probably don't want to make the spec so large that the overall size gets even close to the limit. I think the default max size is 1.5MB, so I thought that what you had in mind when you suggested we should keep the spec under 1MB.

I was thinking that we should just decide on the budget for the spec and then just make the tradeoffs needed to stay within it. And using two thirds of the overall size seemed reasonable. I think that would give room for the metadata fields, even if some new ones are added in the future. It might even allow us room for a limited status if we should find a need for that.

Restricting the number of devices when counters are being used seems like it could be a path forward, although I'm a bit worried about new features all having some impact on the max number of devices. So we should just think about the overall strategy for managing the size as we keep adding features.

pohly replied:
I turned @mortent's validation test into an integration test (pushed to https://github.com/mortent/kubernetes/pull/2). It's useful by itself because we don't have many (any?) tests which actually exercise the v1beta2 endpoints yet.

Luckily, ManagedFields is perfectly harmless for ResourceSlice:
```
        Total size: 1075096
        ManagedFields size: 6791 (1%)
        Spec size: 1060218 (99)%
        
        ManagedFields:
        [
          {
            "manager": "dra.test",
            "operation": "Update",
            "apiVersion": "resource.k8s.io/v1beta2",
            "time": "2025-03-24T08:54:29Z",
            "fieldsType": "FieldsV1",
            "fieldsV1": {
              "f:metadata": {
                "f:annotations": {
                  ".": {},
                  "f:xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0000.com/xxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxxx0000": {},

...
              "f:spec": {
                "f:devices": {},
                "f:driver": {},
                "f:perDeviceNodeSelection": {},
                "f:pool": {
                  "f:generation": {},
                  "f:name": {},
                  "f:resourceSliceCount": {}
                },
                "f:sharedCounters": {}
              }
```

The ManagedFields are mostly about labels and annotations.

I am not exactly sure about the 1 MiB vs. 1.5 MiB limit. In practice (experiment!), 1.5MiB is the hard limit. 1 MiB might be the recommendation?

pohly replied:
> I was thinking that we should just decide on the budget for the spec and then just make the tradeoffs needed to stay within it. And using two thirds of the overall size seemed reasonable. I think that would give room for the metadata fields, even if some new ones are added in the future. It might even allow us room for a limited status if we should find a need for that.

Yes, keeping some some reserve for the status definitely is part of the plan.

So, two options now:
* Allow 32 attributes + quantities, 16 counters per device => maximum size gets even larger.
* Same, but limit the number of devices per slice to 64 if (and only if) counters are used.

pohly replied:
Option 1 with 32 string attributes and 16 counters plus some labels gives us 1.402244568 MiB total size - much too large! With the validation in master the worst case is even more (32 * 32 counters per device).

Option 2 makes it viable with ~half of that.

We could even allow 32 counters again. The Nvidia example already uses 15, so 16 is a bit too close for comfort... Then we have 0.85 MiB.

pohly replied:
But then question becomes: how many device per slice are needed for the Nvidia (and other) use cases?

We can't count on mixins or other compression mechanisms. They are not in 1.33.

mortent replied:
I agree that option 1 isn't an option. Even with the current threshold on the number of devices/attributes/counters we are too close to the limit.

I think reducing the number of devices to 64 seems reasonable. The only thing that might get a bit awkward is that the limit will apply for devices with counters, even when we get mixins. But although mixins should give users the option to describe devices in a more compact way, it will still be possible to make the device definitions as verbose as without them.

Looking at the example of the DGXA100 example in the KEP, it has 26 devices per physical device. So at least a limit of 64 devices should be well below what is needed for the example we know about.

If more devices are needed, I think the path forward would be to allow sharing of SharedCounters between ResourceSlices in the same pool (tracked in https://github.com/kubernetes/kubernetes/issues/130785). That would give users the option of spreading devices across multiple ResourceSlices, even if they pull from the same counters.

mortent replied:
So thinking about this a bit more, I think we want to decide if we want to set limits that strictly makes sure that it is not possible to make a ResourceSlice that exceeds the Kubernetes size limit, or if we want to make it the responsibility of the drivers to make sure they don't extend the limit?

The first option sounds clean and convenient, but it means that even if someone uses mixins to minimize the amount of configuration, they will still be bound by the device limit that is set based on the worst-case scenario. If we make the responsibility of the driver, then it would allow ResourceSlices with a more compact representation to support more devices.

johnbelamaric replied:
Also 1.5Mi is a default for etcd not a fixed value. Although as I recall the behavior when you exceed the limit at least used to be truncation, rather than rejection. Rejection would be much better.

pohly replied:
> Also 1.5Mi is a default for etcd not a fixed value.

It's configurable, but we cannot choose limits in the validation which then don't work with the default size limit.

> Although as I recall the behavior when you exceed the limit at least used to be truncation, rather than rejection. Rejection would be much better.

Too large objects get rejected, I tried that.

> So thinking about this a bit more, I think we want to decide if we want to set limits that strictly makes sure that it is not possible to make a ResourceSlice that exceeds the Kubernetes size limit, or if we want to make it the responsibility of the drivers to make sure they don't extend the limit?

I believe we have to enforce it with validation. If we allow driver authors to make the spec larger, then they might go for something that really uses, say, 1.5 MiB, *because it works now*. Then if we want to add anything in the future like something in the status, we can't, because it would break deployments that used to work.

We could look at the actual encoding during the validation and say "total `spec.Size()` < 1MiB". But that makes validation depend on an implementation detail of the protobuf encoding. That's not good because that exact implementation then effectively gets promoted to API and must remain stable. It's also not precise (validation sees the size of the internal API version, stored version is different).

Here's another option:
- We allow 32 attributes/quantities per device.
- We allow 32 counters per device.
- We allow 128 * 16 counters in devices per slice.

That gives us a maximum size that should be well below 1MibB (haven't tried it). It should match the Nvidia use case (24 devices, 15 counters per device, 25 attributes/capacities).

Then once we have mixins, someone can use those to indirectly define more than 128 * 16 counters if they need that.

johnbelamaric replied:
Yes, I think leaving that flexibility so that people can utilize mixins to increase the device count is important. If counters == 0 then what will be the device limit? Same as now, I assume (no need to reduce in that case).

We might eventually want to shift to parts of resource slices being across the whole pool rather than internally validate in an individual slice, as described in #130785. @pohly is that something that's even possible to change later, from the point of view of validation? Maybe with multiple releases?

mortent replied:
Yeah, I like the idea of defining the limits in such a way that utilizing features that leads to more compact device specs also allows for more devices. Constraining the total number of counters in a slice could be a good way to do this.

I will test out the size with the constraints proposed by @pohly and we can see whether we end up with a worst-case size that we are comfortable with. I will also pick the updates to the tests into the PR branch.

pohly replied:
> If counters == 0 then what will be the device limit? Same as now, I assume (no need to reduce in that case).

Yes, 128 devices.

> We might eventually want to shift to parts of resource slices being across the whole pool rather than internally validate in an individual slice, as described in https://github.com/kubernetes/kubernetes/issues/130785. @pohly is that something that's even possible to change later, from the point of view of validation? Maybe with multiple releases?

Yes. You need to relax validation in release X for existing objects, then in X+1 allow the relaxed validation also for creation and updates.

pohly replied:
BTW, besides overall size there is another reason for these limits: preventing the creation of objects that Kubernetes cannot handle efficiently.

For partitionable devices we don't know yet how scalable the algorithm will be. We will have to do benchmarking to find out where the painpoints are, how much https://github.com/kubernetes/kubernetes/issues/130874 can help, and then what the number of counters are that realistically can be supported.

mortent replied:
So I did some additional checks on this, and I realized that I had made an incorrect assumption in that `ConsumesCounters` in each device would be more expensive than `Attributes`. This is not correct, although it might be a bug ref https://github.com/kubernetes/kubernetes/pull/129970#issuecomment-2749278389. So I've tried the test with using the full capacity/attributes/counters quota (32) for attributes.

But what I find now is that a worst case ResourceSliceSpec without counters, taints or per-device-node-selection is already at almost 700KB. And just adding taints gets it up to 1.1MB.

So I don't think there is any way to add support for counters while keeping the 128 device limit. Even keeping that limit when adding support for taints seems questionable.

So unless I'm missing something here, I think our best option is to reduce the device count to 64 if any of the new features (counters, taints, per-device-node-selection) is used in the ResourceSlice. Although we need to check the max size when all the features are being used even at that limited device count.

johnbelamaric replied:
I think that would be OK. That, or make the validation on the count of total items populated across all devices, rather than on individual fields. Which is a little hard for users, since it's not easily tracked field-by-field, but would offer flexibility. 

We have a solution for this (mixins + #130785); we just have to make sure the interim solution is useable and can cleanly integrate that solution.

pohly replied:
If we limit it to "64 if any of the new features (counters, taints, per-device-node-selection) is used in the ResourceSlice", then that's a decision that we cannot easily relax within a single release.

My preference is still what I proposed in https://github.com/kubernetes/kubernetes/pull/129970#discussion_r2010401343: we allow 128 * 16 counters in devices per slice and 32 counters per device.

Regarding "just adding taints gets it up to 1.1MB": we could limit the number of taints per device from 8 to 4. I think that should still be enough. It if isn't, then there's DeviceTaintRule as an escape hatch. In contrast to counters which are perfectly deterministic (= part of the driver design), taints might be less predictable. We don't want to end up in a situation where a counter can only taint some devices but not others because of some overall count.

mortent replied:
So I agree that we should try hard to keep the limit on the number of devices to 128. It just sets us up better for when mixins comes in 1.34.

We could change the number of taints from 8 to 4. Sounds like we want to avoid a ResourceSlice-wide limit on those.

I've tried a scenario with:
* 128 devices
* 32 annotations per device (no capacity since annotations are more expensive)
* 16 counters per device
* 4 taints per device

This results in a ResourceSliceSpec of size 1,241,854. I think that is too high. So we probably need to tighten the limits a bit.

One option is to combine the limit across annotations/capacity and counters. So maybe support an average of 32 annotations/capacity/counters per device? In the worst-case example, these would all be annotations and the size would fall to 944,894.

Another option would be to only support an average of 24 annotations/capacity across all devices. That reduces the size to 1,069,822

@pohly Do you have a target size in mind? I was thinking somewhere around 700k, as that would give us some room to add new features like mixins. But it does require us to make some tradeoffs now.

pohly replied:
> 24 annotations/capacity across all devices

The current limit is 32. We can't reduce that.

What is wrong with [my proposal](https://github.com/kubernetes/kubernetes/pull/129970#discussion_r2010401343)? You seem to keep looking for other alternatives.

pohly replied:
> We could change the number of taints from 8 to 4.

Yes, let's do that. Your post may have overlapped with mine. I was proposing the same.

pohly replied:
Sorry, I missed that this was in direct response to my proposal.

What if we reduce it to 64 * 16 counters in total? That still works for the Nvidia case (it has 26 devices with 15 counters each).

mortent replied:
Reducing the number of counters allowed in a ResourceSlice helps a little bit, but it still has a worst-case size of 1,093,374 bytes.

Our main challenge is that the attributes are very expensive. 32 * 128 attributes alone takes 688,128 bytes. So changing the number of counters or taints helps a little bit, but to really make a difference it seems like we need to either reduce the number of attributes or the number of devices. And we don't really like either of those options.

pohly replied:
> Reducing the number of counters allowed in a ResourceSlice helps a little bit, but it still has a worst-case size of 1,093,374 bytes.

I think that's good enough. It's not going to be reached in practice and leaves us enough space for status and additional extensions, if we are careful.

As you noted, it's mostly the attributes which can be large. Perhaps we should have done something like "total number of attributes" like it's now proposed for counters, but that was before we had line-of-sight towards mixins as a way to provide attributes more efficiently. Anyway, it's too late now.

mortent replied:
Sounds good. I agree that the worst-case will not be used in practice, and I think normal annotations will be far smaller than the worst-case.

I will update the PR with the additional validation.

mortent replied:
These changes are in a new commit and is ready for review.

jgehrcke replied:
Thanks for all the great work here and the thoughtful discussion. I have a bad conscience because we didn't engage anymore today. We were debugging a last-minute problem with our (NVIDIA's) DRA driver before we want to push out a release still this week.

pohly replied:
I think this discussion has been resolved, using 1024 (= 64 * 16) as the total number of counters in devices.

https://github.com/kubernetes/kubernetes/pull/129970/commits/935dbb92d56a9be9da89a5443342ea59237b899b#r2008406251 may still be open.

pohly replied:
Well, almost:

```suggestion
	// The maximum number of attributes and capacities combined is 32.
```

dims commented:
Is the plan to switch to `Default: true` here?
https://github.com/kubernetes/kubernetes/blob/37afe38abf4edf76b0a3fb820e4e963263dda049/pkg/features/kube_features.go#L1146

pohly commented:
> Is the plan to switch to Default: true here?

No. We can't because the API group has to remain off by default.

thockin commented:
FTR: This still needs LGTM and is currently on hold.

mortent commented:
@cici37 @johnbelamaric When we introduced counters, is it intentional that the requirements for the name of a counter differs from attributes and capacity? For attributes and capacity, we use the `QualifiedName` for this, while for counters we just use a string.

As part of this PR, I realized that there is an inconsistency in the validation for the counters maps, both in the CounterSet and in the DeviceCounterConsumption. In the call to `validateMap`, we specify [`attributeAndCapacityMaxKeyLength`](https://github.com/kubernetes/kubernetes/blob/62555cadc741a8aef31acab05e9ed60f897edb1b/pkg/apis/resource/validation/validation.go#L52) as the max length of the key (which is the same as we use for capacity and attributes and is 96 characters), but the validation function for the key ([validateCounterName](https://github.com/kubernetes/kubernetes/blob/62555cadc741a8aef31acab05e9ed60f897edb1b/pkg/apis/resource/validation/validation.go#L49)) uses `ValidateDNS1123Label` which limits the length to 63 characters.

I assume the intention here is that we should allow the same format on the counters as we do on the capacity and the attributes. And if so, is there a reason why we decided not to use the `QualifiedName` type as the key?

johnbelamaric commented:
> @cici37 @johnbelamaric When we introduced counters, is it intentional that the requirements for the name of a counter differs from attributes and capacity? For attributes and capacity, we use the `QualifiedName` for this, while for counters we just use a string.

Yes. Counters are internal only, whereas attributes and capacities are external and can be shared between drivers. There is no need for qualified names for counters, they never leave the resource slice (until maybe eventually scoping to the resource pool, but that's still all internal to the same driver with no user interaction).

mortent commented:
> Yes. Counters are internal only, whereas attributes and capacities are external and can be shared between drivers. There is no need for qualified names for counters, they never leave the resource slice (until maybe eventually scoping to the resource pool, but that's still all internal to the same driver with no user interaction).

Ack. So for counter names, we validate it according to the DNS1123 label requirements. I've updated the validation checks to reflect this.

pohly commented on pkg/apis/resource/validation/validation_resourceslice_test.go:
```diff
@@ -820,20 +900,104 @@ func createSharedCounters(count int) []resourceapi.CounterSet {
 	return consumeCapacity
 }
+
+func TestResourceSliceSpecSize(t *testing.T) {
```
Shall we perhaps remove this and focus instead on doing this kind of check via the integration test? I think it is a bit more complete over there.

mortent replied:
Agree with that. Just for now the integration tests are failing and I haven't had time to look into that yet. If you happen to know what the issue might be, please let me know.

pohly replied:
The taint key is missing in `testPublishResourceSlices`. One only sees the error when running with at least -v=3 - not ideal, we should add an "inform me about problems with my slices" Go API to the resourceslice controller in 1.34.

I'm a bit puzzled why that surfaces now. `validateDeviceTaints` was called, only the length check was missing.
Ah! I fixed the "all" integration test case such that it now enables the feature. That was missing before, so the faulty taint got dropped. :man_facepalming:

mortent replied:
I think I found the missing taint key, but not sure what needs to happen with the "all" integration test case.

I have updated the `NewMaxResourceSlice` function to reflect the changes and removed the max size unit test.

pohly replied:
I already fixed the "all" configuration in one of the changes that you copied from me into this PR - that is what caused test failure because with the fixed configuration, the apiserver no longer dropped the invalid field and complained instead.

cici37 commented:
> > Yes. Counters are internal only, whereas attributes and capacities are external and can be shared between drivers. There is no need for qualified names for counters, they never leave the resource slice (until maybe eventually scoping to the resource pool, but that's still all internal to the same driver with no user interaction).
> 
> Ack. So for counter names, we validate it according to the DNS1123 label requirements. I've updated the validation checks to reflect this.

Yes, that's why [validateCounterName](https://github.com/kubernetes/kubernetes/blob/8418f228810e4e41c1ac74e4c048c38ae74d227d/pkg/apis/resource/validation/validation.go#L49) got added in validation^^

seans3 commented:
/triage accepted

mortent commented:
Both are flakes.
/test pull-kubernetes-e2e-gce
/test pull-kubernetes-unit

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -265,45 +247,37 @@ type Device struct {
 	// The name of each capacity must be unique in that set.
 	//
-	// The maximum number of attributes and capacities combined is 32.
+	// The maximum number of attributes, capacities and counters combined is 32.
```
```suggestion
	// The maximum number of attributes and capacities combined is 32.
```

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -265,45 +247,37 @@ type Device struct {
+	// The total number of device counter consumption entries
+	// must be <= 32. In addition, the total number in the
+	// entire ResourceSlice must be <= 1024 (for example,
+	// 64 devices with 16 counters each).
```
This resolves https://github.com/kubernetes/kubernetes/pull/129970/files#r2008408292 (adding a link because that thread is soon going to be considered "outdated" by GitHub).

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -351,17 +327,15 @@ type BasicDevice struct {
+	// Counters defines the counters that will be consumed by the device.
 	//
-	// The maximum number of Counters is 32.
+	// The maximum number of attributes, capacities and counters in a device is 32.
```
```suggestion
	// The maximum number counters in a device is 32.
	// In addition, the maximum number of all counters
	// in all devices is 1024 (for example, 64 devices with
	// 16 counters each).
```

pohly commented on pkg/apis/resource/types.go:
```diff
@@ -386,9 +360,17 @@ type Counter struct {
 }
 
-// Limit for the sum of the number of entries in both attributes and capacity.
+// Limit for the sum of the number of entries in attributes, capacity.
```
```suggestion
// Limit for the sum of the number of entries in both attributes and capacity.
```

pohly reviewed (commented):
Still some cleanups pending and https://github.com/kubernetes/kubernetes/pull/129970/files#r2010639071 is unresolved?

mortent commented:
> Still some cleanups pending and https://github.com/kubernetes/kubernetes/pull/129970/files#r2010639071 is unresolved?

I've fixed the comments. I think the issue should be resolved, as our current design does leave sufficient room for both attributes+capacity and counters. I added a more detailed comment on the thread.

pohly reviewed (approved):
/lgtm

Not lifting the hold yet, https://github.com/kubernetes/kubernetes/pull/129970#discussion_r2013650218 may still be open.

We might also want to squash one last time before merging.

mortent committed:
Run make update

mortent committed:
Add e2e tests for resource v1beta2 API

mortent commented on pkg/apis/resource/types.go:
```diff
@@ -452,7 +437,7 @@ type DeviceAttribute struct {
 
 // DeviceTaintsMaxLength is the maximum number of taints per device.
-const DeviceTaintsMaxLength = 8
+const DeviceTaintsMaxLength = 4
```
Fixed.

pohly replied:
/lgtm

johnbelamaric commented:
@klueska is out but I believe we have addressed his concerns.

/approve
/hold cancel

pohly commented:
/retest

Unrelated flake ("failed to add route {Ifindex: 2 Dst: fd00:10:244:1::1/128 Src: fd00:10:244:1::1b6 Gw: <nil> Flags: [] Table: 0 Realm: 0}: invalid argument").
